elastic/kibana
# [Fleet] Don't add extra quotes to YAML strings from manifest files

## Summary

Fixes https://github.com/elastic/kibana/issues/93253

This reverts the changes made in https://github.com/elastic/kibana/pull/91418 , and adds extra validation to the Policy UI in Fleet to guard against invalid YAML string values.

With this change, string default values in YAML files in packages need to be enclosed in quotes like this when they start with the wildcard `*`:
- `'"*"'`, `'"*foo"'` or
- `"\"*\""`, `"\"*foo\""`

When a user tries to enter such a string when creating a policy, it is caught by the form validation.

### Background
Currently, the package registry reads in package metadata from YAML files in packages, and presents it to the Kibana Fleet plugin in JSON format. For an example, see the `policy_templates` section in https://epr.elastic.co/package/apache/0.3.4/ .

If the package contains default values for string properties, the information if these strings were quoted or not gets lost during the translation to JSON, so

```
- name: some_var
  type: text
  default: "*"
```
and 
```
- name: some_var
  type: text
  default: *
```
would both result in the JSON output
```
{
  "name": "some_var",
  "default": "*"
}
```
in the registry output. When this JSON object is turned into YAML again later, the string in question is left unquoted, so it is now
```
default: *
```
in both cases. When that YAML is parsed again, it leads to a parser error because the string starts with a `*`, which for YAML means this should be parsed as an alias.

https://github.com/elastic/kibana/pull/91418 tried to amend this by cleverly adding quotes later in the process for strings which were deemed problematic. This led to other bugs (see https://github.com/elastic/kibana/issues/93253 ) and was clearly not a good solution.

**This change now uses a different approach, and gives the responsibility to provide correctly quoted strings to the package author, and the user. Package authors now must take care to use escaped quotation marks in case they need to be preserved in the final YAML policy. Users are helped by additional form validation which points out when they need to enclose their strings in `""`.**

### How to test this

* Add any integration to any policy. Try to enter a string like `*` or `*foo` in a text field, and verify that you can't save the integration. Play around with it and try to break it.
![image](https://user-images.githubusercontent.com/189073/109962500-8e361580-7ceb-11eb-8b26-5908f9e17552.png)

* Add quotes as requested by the form validation, save the integration, and verify that the operation is successful and the generated policy looks correct.

* Craft a package with a string default value like `'*'` or `"*foo"` and try to add it to a policy. Verify that you see the same error in the form validation as above. (The current APM package in the snapshot registry has this problem.)

* Fix the package so that the string default value now contains escaped quotes, like `'"*"'` or `"\"*foo\""`. Verify that you can add the integration to a policy correctly.

* Verify that you can add the `kafka` integration to a policy, and that the policy contains these lines:
![image](https://user-images.githubusercontent.com/189073/109964083-824b5300-7ced-11eb-8800-361f9b6b4ace.png)

### Checklist

Delete any items that are not applicable to this PR.
